src/marlin_stubs/sdcard/M20-M30_M32-M34.cpp

```diff
@@ -5,18 +5,57 @@
 #include "media.h"
 #include "marlin_vars.hpp"
 
-// M20 - List SD card
-void GcodeSuite::M20() {
-    SERIAL_ECHOLNPGM(MSG_BEGIN_FILE_LIST);
+static long GetFileSize(std::string filename) {
+    struct stat stat_buf;
+    int rc = stat(filename.c_str(), &stat_buf);
+    return rc == 0 ? stat_buf.st_size : -1;
+}
+
+static void printListing(const char *const prepend, const char *const prependLong, bool timestamps, bool longfilenames) {
     DIR *dir;
-    dir = opendir("/usb/");
+    dir = opendir(prepend);
     if (dir != NULL) {
         struct dirent *entry;
         while ((entry = readdir(dir)) != NULL && entry->d_name[0]) {
-            SERIAL_ECHOLN(entry->d_name);
+            std::string dirname(std::string(prepend) + "/" + entry->d_name);
+            std::string longdirname(std::string(prependLong) + "/" + entry->lfn);
+            if (entry->d_type == DT_DIR) {
+                printListing(dirname.c_str(), longdirname.c_str(), timestamps, longfilenames);
```

Comment:
Recursion may quickly lead to stack overflow here.

Reply:
That's how it was handled in the past. I'll come up with a for loop instead.

Reply:
Seems like dir parsing introduces more issues than expected (opendir will hang indefinitely), so I remove parsing subdirectories for this commit.